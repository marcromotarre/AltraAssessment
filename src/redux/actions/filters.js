import * as types from './../constants';

export const setAgeFilter = ({ filterData }) => ({
  type: types.SET_AGE_FILTER,
  payload: {
    filterData,
  },
});

export const setAgeMin = ({ min }) => ({
  type: types.SET_AGE_MIN,
  payload: {
    min,
  },
});

export const setAgeMax = ({ max }) => ({
  type: types.SET_AGE_MAX,
  payload: {
    max,
  },
});

export const setAgeRange = ({ range }) => ({
  type: types.SET_AGE_RANGE,
  payload: {
    range,
  },
});

export const setHeightFilter = ({ filterData }) => ({
  type: types.SET_HEIGHT_FILTER,
  payload: {
    filterData,
  },
});

export const setHeightMin = ({ min }) => ({
  type: types.SET_HEIGHT_MIN,
  payload: {
    min,
  },
});

export const setHeightMax = ({ max }) => ({
  type: types.SET_HEIGHT_MAX,
  payload: {
    max,
  },
});

export const setHeightRange = ({ range }) => ({
  type: types.SET_HEIGHT_RANGE,
  payload: {
    range,
  },
});

export const setWeightFilter = ({ filterData }) => ({
  type: types.SET_WEIGHT_FILTER,
  payload: {
    filterData,
  },
});

export const setWeightMin = ({ min }) => ({
  type: types.SET_WEIGHT_MIN,
  payload: {
    min,
  },
});

export const setWeightMax = ({ max }) => ({
  type: types.SET_WEIGHT_MAX,
  payload: {
    max,
  },
});

export const setWeightRange = ({ range }) => ({
  type: types.SET_WEIGHT_RANGE,
  payload: {
    range,
  },
});

export const addHairColorFilter = ({ color }) => ({
  type: types.ADD_HAIR_COLOR_FILTER,
  payload: {
    color,
  },
});

export const deleteHairColorFilter = ({ color }) => ({
  type: types.DELETE_HAIR_COLOR_FILTER,
  payload: {
    color,
  },
});

export const clearHairColorsFilters = () => ({
  type: types.CLEAR_HAIR_COLORS_FILTER,
});
