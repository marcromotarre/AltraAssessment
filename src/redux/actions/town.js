import * as types from './../constants';

export function fetchTownPending() {
  return {
    type: types.FETCH_TOWN_PENDING,
  };
}

export function fetchTownSuccess(response) {
  return {
    type: types.FETCH_TOWN_SUCCESS,
    payload: {
      response,
    },
  };
}

export function fetchTownError(error) {
  return {
    type: types.FETCH_TOWN_ERROR,
    payload: {
      error,
    },
  };
}
