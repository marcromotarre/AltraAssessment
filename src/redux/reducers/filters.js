import * as types from './../constants';

export const initialState = {
  age: {
    min: '',
    max: '',
    range: true,
  },
  weight: {
    min: '',
    max: '',
    range: true,
  },
  height: {
    min: '',
    max: '',
    range: true,
  },
  hair_colors: [],
};

export function reducer(state = initialState, { type, payload }) {
  switch (type) {
    case types.SET_AGE_FILTER:
      return {
        ...state,
        age: payload.filterData,
      };
    case types.SET_AGE_MIN:
      return {
        ...state,
        age: { ...state.age, min: payload.min },
      };
    case types.SET_AGE_MAX:
      return {
        ...state,
        age: { ...state.age, max: payload.max },
      };
    case types.SET_AGE_RANGE:
      return {
        ...state,
        age: { ...state.age, range: payload.range },
      };
    case types.SET_HEIGHT_FILTER:
      return {
        ...state,
        height: payload.filterData,
      };
    case types.SET_HEIGHT_MIN:
      return {
        ...state,
        height: { ...state.height, min: payload.min },
      };
    case types.SET_HEIGHT_MAX:
      return {
        ...state,
        height: { ...state.height, max: payload.max },
      };
    case types.SET_HEIGHT_RANGE:
      return {
        ...state,
        height: { ...state.height, range: payload.range },
      };
    case types.SET_WEIGHT_FILTER:
      return {
        ...state,
        weight: payload.filterData,
      };
    case types.SET_WEIGHT_MIN:
      return {
        ...state,
        weight: { ...state.weight, min: payload.min },
      };
    case types.SET_WEIGHT_MAX:
      return {
        ...state,
        weight: { ...state.weight, max: payload.max },
      };
    case types.SET_WEIGHT_RANGE:
      return {
        ...state,
        weight: { ...state.weight, range: payload.range },
      };
    case types.ADD_HAIR_COLOR_FILTER:
      return {
        ...state,
        hair_colors: !state.hair_colors.includes(payload.color)
          ? [...state.hair_colors, payload.color]
          : [...state.hair_colors],
      };

    case types.DELETE_HAIR_COLOR_FILTER:
      return {
        ...state,
        hair_colors: state.hair_colors.filter(
          (color) => color !== payload.color
        ),
      };

    case types.CLEAR_HAIR_COLORS_FILTER:
      return {
        ...state,
        hair_colors: [],
      };

    default:
      return {
        ...state,
      };
  }
}
