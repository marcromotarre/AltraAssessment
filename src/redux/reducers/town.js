import * as types from './../constants';
import {
  getTownsNames,
  getUsersByTown,
  getUniqueHairColorsByTown,
} from '../../utils/town';

export const initialState = {
  pending: false,
  townName: '',
  users: [],
  uniqueHairColors: [],
  error: null,
};

export function reducer(state = initialState, { type, payload }) {
  switch (type) {
    case types.FETCH_TOWN_PENDING:
      return {
        ...state,
        pending: true,
      };
    case types.FETCH_TOWN_SUCCESS:
      const townName = getTownsNames(payload.response)[0];
      return {
        ...state,
        pending: false,
        townName,
        users: getUsersByTown(payload.response, townName),
        uniqueHairColors: getUniqueHairColorsByTown(payload.response, townName),
      };
    case types.FETCH_TOWN_ERROR:
      return {
        ...state,
        pending: false,
        error: payload.error,
      };

    default:
      return {
        ...state,
      };
  }
}
