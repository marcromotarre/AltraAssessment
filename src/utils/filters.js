export const isMinMaxFilterApplied = (filter) =>
  filter.min === '' && filter.max === '' ? false : true;

export const isArrayFilterApplied = (filter) =>
  filter.length > 0 ? true : false;

export const getFilterMinMaxString = (data) => {
  if (data.range && data.min !== '' && data.max !== '') {
    return `${data.min} - ${data.max}`;
  } else if (data.min !== '') {
    return `${data.min}`;
  } else if (data.max !== '') {
    return `${data.max}`;
  } else {
    return '';
  }
};

export const isAgeOrWeightOrHeightFilterApplied = ({ age, height, weight }) => {
  return isMinMaxFilterApplied(age) ||
    isMinMaxFilterApplied(height) ||
    isMinMaxFilterApplied(weight)
    ? true
    : false;
};

export const isAnyFilterApplied = ({ age, height, weight, hair_colors }) => {
  return isAgeOrWeightOrHeightFilterApplied({ age, height, weight }) ||
    isArrayFilterApplied(hair_colors)
    ? true
    : false;
};

export const getUsersByName = (users, name) => {
  return users.filter((user) =>
    user.name.toLowerCase().includes(name.toLowerCase())
  );
};

export const getUsersByProfession = (users, profession) => {
  return users.filter((user) => user.professions.includes(profession));
};

export const getUniqueHairColors = (users) => {
  return users
    .map((user) => user.hair_color)
    .filter((value, index, self) => self.indexOf(value) === index);
};

export const getUsersByMinMaxFilter = (users, type, filter) => {
  if (!isMinMaxFilterApplied(filter)) {
    return [...users];
  } else {
    if (filter.min !== '' && filter.max !== '' && filter.range) {
      return users.filter((user) => {
        return (
          user[type] >= parseFloat(filter.min) &&
          user[type] <= parseFloat(filter.max)
        );
      });
    } else if (
      (filter.min !== '' && filter.max !== '' && !filter.range) ||
      filter.min !== ''
    ) {
      return users.filter((user) => {
        return Math.round(user[type]) === parseFloat(filter.min);
      });
    } else {
      return users.filter((user) => {
        return Math.round(user[type]) === parseFloat(filter.max);
      });
    }
  }
};

export const getUsersByArrayFilter = (users, type, filter) => {
  if (!isArrayFilterApplied(filter)) {
    return [...users];
  } else {
    return users.filter((user) => {
      return filter.includes(user[type]);
    });
  }
};

export const getUsersByFilters = ({
  users = [],
  age = { min: '', max: '', range: false },
  height = { min: '', max: '', range: false },
  weight = { min: '', max: '', range: false },
  hair_colors = [],
}) => {
  return getUsersByArrayFilter(
    getUsersByMinMaxFilter(
      getUsersByMinMaxFilter(
        getUsersByMinMaxFilter(users, 'age', age),
        'weight',
        weight
      ),
      'height',
      height
    ),
    'hair_color',
    hair_colors
  );
};
