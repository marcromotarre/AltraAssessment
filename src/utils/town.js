import { getUniqueHairColors } from './filters';

export const getTownsNames = (data) => Object.keys(data);

export const getUsersByTown = (data, townName) => data[townName];

export const getUniqueHairColorsByTown = (data, townName) =>
  getUniqueHairColors(data[townName]);
