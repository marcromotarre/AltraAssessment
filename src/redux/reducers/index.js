import { combineReducers } from 'redux';
import { reducer as town } from './town';
import { reducer as filters } from './filters';

export default combineReducers({
  town,
  filters,
});
