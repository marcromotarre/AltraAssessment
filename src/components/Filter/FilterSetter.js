import React from 'react';
import styled from 'styled-components';
import { useSelector, useDispatch, connect } from 'react-redux';

import CloseRedIcon from './../../svg/x.svg';

import * as actions from '../../redux/actions/filters';

import { isArrayFilterApplied } from './../../utils/filters';

import {
  getMinUserValueByField,
  getMaxUserValueByField,
} from './../../utils/limits';

import FilterMinMax from './FilterMinMax.js';
import ColoredHair from '../User/ColoredHair';
import FilterMultipleSelection from './FilterMultipleSelection';
import { colors } from '../../data/colors';

const FilterSetterContainer = styled.section`
  align-items: stretch;
  display: flex;
  flex-flow: column;
  height: 100%;
  flex: auto;
  background: #f7f8fd;
  width: 100%;
`;

const NumberFilteredUsers = styled.section`
  font-family: 'HelveticaNeue-Thin';
  color: #000;
  font-size: 18px;
  justify-self: center;
  align-self: center;
`;

const Filters = styled.section`
  width: 100%;
  display: grid;
  grid-template-columns: 100%;
  row-gap: 10px;
`;

const Filter = styled.div`
  display: grid;
  grid-template-columns: 80% 20%;
  > * {
    justify-self: center;
    align-self: center;
  }
`;

const FilterSetter = ({ filteredUsersNumber }) => {
  const dispatch = useDispatch();
  const { users, uniqueHairColors } = useSelector((state) => state.town);
  const { age, height, weight, hair_colors } = useSelector(
    (state) => state.filters
  );

  const clearAgeFilter = () =>
    dispatch(
      actions.setAgeFilter({
        filterData: { min: '', max: '', range: age.range },
      })
    );

  const ageMinValueChanged = (value) =>
    dispatch(actions.setAgeMin({ min: value }));

  const ageMaxValueChanged = (value) =>
    dispatch(actions.setAgeMax({ max: value }));

  const ageRangeValueChanged = (value) =>
    dispatch(actions.setAgeRange({ range: value }));

  const clearHeightFilter = () =>
    dispatch(
      actions.setHeightFilter({
        filterData: { min: '', max: '', range: height.range },
      })
    );
  const heightMinValueChanged = (value) =>
    dispatch(actions.setHeightMin({ min: value }));

  const heightMaxValueChanged = (value) =>
    dispatch(actions.setHeightMax({ max: value }));

  const heightRangeValueChanged = (value) =>
    dispatch(actions.setHeightRange({ range: value }));

  const clearWeightFilter = () =>
    dispatch(
      actions.setWeightFilter({
        filterData: { min: '', max: '', range: weight.range },
      })
    );

  const weightMinValueChanged = (value) =>
    dispatch(actions.setWeightMin({ min: value }));

  const weightMaxValueChanged = (value) =>
    dispatch(actions.setWeightMax({ max: value }));

  const weightRangeValueChanged = (value) =>
    dispatch(actions.setWeightRange({ range: value }));

  const handleOnClickHairColor = (index) => {
    const uniqueHairColor = uniqueHairColors[index];
    const isSelected = hair_colors.includes(uniqueHairColor);
    if (isSelected) {
      dispatch(actions.deleteHairColorFilter({ color: uniqueHairColor }));
    } else {
      dispatch(actions.addHairColorFilter({ color: uniqueHairColor }));
    }
  };

  const clearHairColorsFilter = () =>
    dispatch(actions.clearHairColorsFilters());

  return (
    <FilterSetterContainer data-testid="FilterSetter">
      <NumberFilteredUsers>{`( ${filteredUsersNumber} )`}</NumberFilteredUsers>
      <Filters>
        <Filter>
          <FilterMinMax
            title="Age"
            minInputValue={age.min}
            maxInputValue={age.max}
            showTo={age.range}
            minLimit={getMinUserValueByField(users, 'age')}
            maxLimit={getMaxUserValueByField(users, 'age')}
            minChanged={ageMinValueChanged}
            maxChanged={ageMaxValueChanged}
            showToChanged={ageRangeValueChanged}
          />
          {(age.min !== '' || age.max !== '') && (
            <button data-testid="clear-age-icon" onClick={clearAgeFilter}>
              <CloseRedIcon width={14} />
            </button>
          )}
        </Filter>

        <Filter>
          <FilterMinMax
            title="Height"
            minInputValue={height.min}
            maxInputValue={height.max}
            showTo={height.range}
            minLimit={getMinUserValueByField(users, 'height')}
            maxLimit={getMaxUserValueByField(users, 'height')}
            minChanged={heightMinValueChanged}
            maxChanged={heightMaxValueChanged}
            showToChanged={heightRangeValueChanged}
          />
          {(height.min !== '' || height.max !== '') && (
            <button data-testid="clear-height-icon" onClick={clearHeightFilter}>
              <CloseRedIcon width={14} />
            </button>
          )}
        </Filter>

        <Filter>
          <FilterMinMax
            title="Weight"
            minInputValue={weight.min}
            maxInputValue={weight.max}
            showTo={weight.range}
            minLimit={getMinUserValueByField(users, 'weight')}
            maxLimit={getMaxUserValueByField(users, 'weight')}
            minChanged={weightMinValueChanged}
            maxChanged={weightMaxValueChanged}
            showToChanged={weightRangeValueChanged}
          />
          {(weight.min !== '' || weight.max !== '') && (
            <button data-testid="clear-weight-icon" onClick={clearWeightFilter}>
              <CloseRedIcon width={14} />
            </button>
          )}
        </Filter>

        <Filter>
          <FilterMultipleSelection
            title="Hair Color"
            clickOnElement={handleOnClickHairColor}
          >
            {uniqueHairColors.map((uniqueHairColor) => {
              const isSelected = hair_colors.includes(uniqueHairColor);
              return (
                <ColoredHair
                  key={uniqueHairColor}
                  strokecolor={
                    isSelected ? colors.Transparent : uniqueHairColor
                  }
                  color={isSelected ? uniqueHairColor : colors.Transparent}
                  width={14.28}
                />
              );
            })}
          </FilterMultipleSelection>
          {isArrayFilterApplied(hair_colors) && (
            <button
              data-testid="clear-colors-icon"
              onClick={clearHairColorsFilter}
            >
              <CloseRedIcon width={14} />
            </button>
          )}
        </Filter>
      </Filters>
    </FilterSetterContainer>
  );
};

export default FilterSetter;
