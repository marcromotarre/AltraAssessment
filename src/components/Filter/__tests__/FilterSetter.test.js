import React from 'react';
import { fireEvent } from '@testing-library/react';
import FilterSetter from '../FilterSetter';
import { initialState as filtersInitialState } from './../../../redux/reducers/filters';
import { initialState as townInitialState } from './../../../redux/reducers/town';

import { renderRedux } from './../../../utils/test';
/** this component should:
 *  - allow user to activate and desactivate filters
 *      - age: min max filter
 *          - clear click on clean icon (only visible when filter is not empty)
 *      - height: min max filter
 *          - clear click on clean icon (only visible when filter is not empty)
 *      - weight: min max filter
 *          - clear click on clean icon (only visible when filter is not empty)
 *      - hair_colors: array filter
 *          - clear click on clean icon (only visible when filter is not empty)
 *      - see how many users pass the filters
 *
 */

const numericEvent34 = {
  target: {
    value: '34',
  },
};

const numericEvent50 = {
  target: {
    value: '50',
  },
};

const initialState = {
  filters: filtersInitialState,
  town: townInitialState,
};

describe('FilterSetter Component', () => {
  it('FilterSetter Snapshot', () => {
    expect(
      renderRedux(<FilterSetter filteredUsersNumber={1000} />, {
        initialState: initialState,
      }).baseElement
    ).toMatchSnapshot();
  });
  it('age min value changes when change filter min input', () => {
    const { getByTestId } = renderRedux(
      <FilterSetter filteredUsersNumber={1000} />,
      {
        initialState: initialState,
      }
    );

    const ageMinInput = getByTestId('Age-filterminmax-min-input');
    expect(ageMinInput.value).toBe('');
    fireEvent.change(ageMinInput, numericEvent34);
    expect(ageMinInput.value).toBe('34');
  });

  it('age max value changes when change filter max input', () => {
    const { getByTestId } = renderRedux(
      <FilterSetter filteredUsersNumber={1000} />,
      {
        initialState: initialState,
      }
    );
    const ageMaxInput = getByTestId('Age-filterminmax-max-input');
    expect(ageMaxInput.value).toBe('');
    fireEvent.change(ageMaxInput, numericEvent50);
    expect(ageMaxInput.value).toBe('50');
  });

  it('age range click', () => {
    const { queryByTestId, getByTestId } = renderRedux(
      <FilterSetter filteredUsersNumber={1000} />,
      {
        initialState: initialState,
      }
    );

    const ageShowHideButton = getByTestId('Age-filterminmax-show-hide-button');
    expect(queryByTestId('Age-filterminmax-min-input')).toBeTruthy();
    expect(queryByTestId('Age-filterminmax-max-input')).toBeTruthy();
    fireEvent.click(ageShowHideButton);
    expect(queryByTestId('Age-filterminmax-min-input')).toBeTruthy();
    expect(queryByTestId('Age-filterminmax-max-input')).toBeFalsy();
    fireEvent.click(ageShowHideButton);
    expect(queryByTestId('Age-filterminmax-min-input')).toBeTruthy();
    expect(queryByTestId('Age-filterminmax-max-input')).toBeTruthy();
  });

  it('age clear icon click', () => {
    const { getByTestId } = renderRedux(
      <FilterSetter filteredUsersNumber={1000} />,
      {
        initialState: {
          ...initialState,
          filters: {
            ...initialState.filters,
            age: { min: '34', max: '50', range: true },
          },
        },
      }
    );
    const ageMinInput = getByTestId('Age-filterminmax-min-input');
    const ageMaxInput = getByTestId('Age-filterminmax-max-input');

    expect(ageMinInput.value).toBe('34');
    expect(ageMaxInput.value).toBe('50');

    const clearAgeIcon = getByTestId('clear-age-icon');
    fireEvent.click(clearAgeIcon);
    expect(ageMinInput.value).toBe('');
    expect(ageMaxInput.value).toBe('');
  });

  it('height min value changes when change filter min input', () => {
    const { getByTestId } = renderRedux(
      <FilterSetter filteredUsersNumber={1000} />,
      {
        initialState: initialState,
      }
    );
    const heightMinInput = getByTestId('Height-filterminmax-min-input');
    expect(heightMinInput.value).toBe('');
    fireEvent.change(heightMinInput, numericEvent34);
    expect(heightMinInput.value).toBe('34');
  });

  it('height max value changes when change filter min input', () => {
    const { getByTestId } = renderRedux(
      <FilterSetter filteredUsersNumber={1000} />,
      {
        initialState: initialState,
      }
    );
    const heightMaxInput = getByTestId('Height-filterminmax-max-input');
    expect(heightMaxInput.value).toBe('');
    fireEvent.change(heightMaxInput, numericEvent50);
    expect(heightMaxInput.value).toBe('50');
  });

  it('height range click', () => {
    const { queryByTestId, getByTestId } = renderRedux(
      <FilterSetter filteredUsersNumber={1000} />,
      {
        initialState: initialState,
      }
    );

    const heightShowHideButton = getByTestId(
      'Height-filterminmax-show-hide-button'
    );
    expect(queryByTestId('Height-filterminmax-min-input')).toBeTruthy();
    expect(queryByTestId('Height-filterminmax-max-input')).toBeTruthy();
    fireEvent.click(heightShowHideButton);
    expect(queryByTestId('Height-filterminmax-min-input')).toBeTruthy();
    expect(queryByTestId('Height-filterminmax-max-input')).toBeFalsy();
    fireEvent.click(heightShowHideButton);
    expect(queryByTestId('Height-filterminmax-min-input')).toBeTruthy();
    expect(queryByTestId('Height-filterminmax-max-input')).toBeTruthy();
  });

  it('height clear icon click', () => {
    const { getByTestId } = renderRedux(
      <FilterSetter filteredUsersNumber={1000} />,
      {
        initialState: {
          ...initialState,
          filters: {
            ...initialState.filters,
            height: { min: '34', max: '50', range: true },
          },
        },
      }
    );
    const weightMinInput = getByTestId('Height-filterminmax-min-input');
    const weightMaxInput = getByTestId('Height-filterminmax-max-input');

    expect(weightMinInput.value).toBe('34');
    expect(weightMaxInput.value).toBe('50');

    const clearAgeIcon = getByTestId('clear-height-icon');
    fireEvent.click(clearAgeIcon);
    expect(weightMinInput.value).toBe('');
    expect(weightMaxInput.value).toBe('');
  });

  it('weight min value changes when change filter min input', () => {
    const { getByTestId } = renderRedux(
      <FilterSetter filteredUsersNumber={1000} />,
      {
        initialState: initialState,
      }
    );
    const weightMinInput = getByTestId('Weight-filterminmax-min-input');
    expect(weightMinInput.value).toBe('');
    fireEvent.change(weightMinInput, numericEvent34);
    expect(weightMinInput.value).toBe('34');
  });

  it('weight max value changes when change filter min input', () => {
    const { getByTestId } = renderRedux(
      <FilterSetter filteredUsersNumber={1000} />,
      {
        initialState: initialState,
      }
    );
    const weightMaxInput = getByTestId('Weight-filterminmax-max-input');
    expect(weightMaxInput.value).toBe('');
    fireEvent.change(weightMaxInput, numericEvent50);
    expect(weightMaxInput.value).toBe('50');
  });

  it('weight range click', () => {
    const { queryByTestId, getByTestId } = renderRedux(
      <FilterSetter filteredUsersNumber={1000} />,
      {
        initialState: initialState,
      }
    );

    const weightShowHideButton = getByTestId(
      'Weight-filterminmax-show-hide-button'
    );
    expect(queryByTestId('Weight-filterminmax-min-input')).toBeTruthy();
    expect(queryByTestId('Weight-filterminmax-max-input')).toBeTruthy();
    fireEvent.click(weightShowHideButton);
    expect(queryByTestId('Weight-filterminmax-min-input')).toBeTruthy();
    expect(queryByTestId('Weight-filterminmax-max-input')).toBeFalsy();
    fireEvent.click(weightShowHideButton);
    expect(queryByTestId('Weight-filterminmax-min-input')).toBeTruthy();
    expect(queryByTestId('Weight-filterminmax-max-input')).toBeTruthy();
  });

  it('weight clear icon click', () => {
    const { getByTestId } = renderRedux(
      <FilterSetter filteredUsersNumber={1000} />,
      {
        initialState: {
          ...initialState,
          filters: {
            ...initialState.filters,
            weight: { min: '34', max: '50', range: true },
          },
        },
      }
    );
    const weightMinInput = getByTestId('Weight-filterminmax-min-input');
    const weightMaxInput = getByTestId('Weight-filterminmax-max-input');

    expect(weightMinInput.value).toBe('34');
    expect(weightMaxInput.value).toBe('50');

    const clearAgeIcon = getByTestId('clear-weight-icon');
    fireEvent.click(clearAgeIcon);
    expect(weightMinInput.value).toBe('');
    expect(weightMaxInput.value).toBe('');
  });

  it('hair selected onClick', () => {
    const { getByTestId } = renderRedux(
      <FilterSetter filteredUsersNumber={1000} />,
      {
        initialState: {
          ...initialState,
          town: {
            ...initialState.town,
            uniqueHairColors: ['Red', 'Green', 'Pink'],
          },
        },
      }
    );
    const hair0 = getByTestId('Hair Color-button-0');
    const hair1 = getByTestId('Hair Color-button-1');
    const hair2 = getByTestId('Hair Color-button-2');

    expect(hair0.querySelector('svg').getAttribute('color')).toBe(
      'rgba(0,0,0,0)'
    );
    expect(hair1.querySelector('svg').getAttribute('color')).toBe(
      'rgba(0,0,0,0)'
    );
    expect(hair2.querySelector('svg').getAttribute('color')).toBe(
      'rgba(0,0,0,0)'
    );

    fireEvent.click(hair0);
    expect(hair0.querySelector('svg').getAttribute('color')).toBe('Red');
    fireEvent.click(hair0);
    expect(hair0.querySelector('svg').getAttribute('color')).toBe(
      'rgba(0,0,0,0)'
    );
  });

  it('hair color clear icon click', () => {
    const { getByTestId } = renderRedux(
      <FilterSetter filteredUsersNumber={1000} />,
      {
        initialState: {
          ...initialState,
          filters: {
            ...initialState.filters,
            hair_colors: ['Pink', 'Red'],
          },
          town: {
            ...initialState.town,
            uniqueHairColors: ['Red', 'Green', 'Pink'],
          },
        },
      }
    );
    const hair0 = getByTestId('Hair Color-button-0');
    const hair1 = getByTestId('Hair Color-button-1');
    const hair2 = getByTestId('Hair Color-button-2');

    expect(hair0.querySelector('svg').getAttribute('color')).toBe('Red');
    expect(hair1.querySelector('svg').getAttribute('color')).toBe(
      'rgba(0,0,0,0)'
    );
    expect(hair2.querySelector('svg').getAttribute('color')).toBe('Pink');

    const clearColorsIcon = getByTestId('clear-colors-icon');
    fireEvent.click(clearColorsIcon);

    expect(hair0.querySelector('svg').getAttribute('color')).toBe(
      'rgba(0,0,0,0)'
    );
    expect(hair1.querySelector('svg').getAttribute('color')).toBe(
      'rgba(0,0,0,0)'
    );
    expect(hair2.querySelector('svg').getAttribute('color')).toBe(
      'rgba(0,0,0,0)'
    );
  });
});
