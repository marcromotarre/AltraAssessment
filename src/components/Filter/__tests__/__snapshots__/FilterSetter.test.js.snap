// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterSetter Component FilterSetter Snapshot 1`] = `
<body>
  <div>
    <section
      class="sc-fzplWN hCinyB"
      data-testid="FilterSetter"
    >
      <section
        class="sc-fznyAO epeGvV"
      >
        ( 1000 )
      </section>
      <section
        class="sc-fznKkj rsgQJ"
      >
        <div
          class="sc-fznZeY jHoiIA"
        >
          <section
            class="sc-AxjAm cUObhj"
          >
            <label
              class="sc-AxirZ pTpRo"
            >
              Age
              :
            </label>
            <input
              class="sc-AxiKw kJgCdA"
              data-testid="Age-filterminmax-min-input"
              value=""
            />
            <label
              class="sc-AxhCb iAbugP"
            >
              min: 
              Infinity
            </label>
            <label
              class="sc-AxhUy cUdIeG"
            >
              to
            </label>
            <input
              class="sc-AxgMl eFELnN"
              data-testid="Age-filterminmax-max-input"
              value=""
            />
            <label
              class="sc-AxheI cUaqWs"
            >
              max: 
              -Infinity
            </label>
            <button
              class="sc-Axmtr fdTANV"
              data-testid="Age-filterminmax-show-hide-button"
            >
              <svg
                width="11.43"
              />
            </button>
          </section>
        </div>
        <div
          class="sc-fznZeY jHoiIA"
        >
          <section
            class="sc-AxjAm cUObhj"
          >
            <label
              class="sc-AxirZ pTpRo"
            >
              Height
              :
            </label>
            <input
              class="sc-AxiKw kJgCdA"
              data-testid="Height-filterminmax-min-input"
              value=""
            />
            <label
              class="sc-AxhCb iAbugP"
            >
              min: 
              Infinity
            </label>
            <label
              class="sc-AxhUy cUdIeG"
            >
              to
            </label>
            <input
              class="sc-AxgMl eFELnN"
              data-testid="Height-filterminmax-max-input"
              value=""
            />
            <label
              class="sc-AxheI cUaqWs"
            >
              max: 
              -Infinity
            </label>
            <button
              class="sc-Axmtr fdTANV"
              data-testid="Height-filterminmax-show-hide-button"
            >
              <svg
                width="11.43"
              />
            </button>
          </section>
        </div>
        <div
          class="sc-fznZeY jHoiIA"
        >
          <section
            class="sc-AxjAm cUObhj"
          >
            <label
              class="sc-AxirZ pTpRo"
            >
              Weight
              :
            </label>
            <input
              class="sc-AxiKw kJgCdA"
              data-testid="Weight-filterminmax-min-input"
              value=""
            />
            <label
              class="sc-AxhCb iAbugP"
            >
              min: 
              Infinity
            </label>
            <label
              class="sc-AxhUy cUdIeG"
            >
              to
            </label>
            <input
              class="sc-AxgMl eFELnN"
              data-testid="Weight-filterminmax-max-input"
              value=""
            />
            <label
              class="sc-AxheI cUaqWs"
            >
              max: 
              -Infinity
            </label>
            <button
              class="sc-Axmtr fdTANV"
              data-testid="Weight-filterminmax-show-hide-button"
            >
              <svg
                width="11.43"
              />
            </button>
          </section>
        </div>
        <div
          class="sc-fznZeY jHoiIA"
        >
          <section
            class="sc-fzozJi cTUUNl"
          >
            <label
              class="sc-fzoLsD nKiEE"
            >
              Hair Color
              :
            </label>
            <section
              class="sc-fzpans fNiaj"
            />
          </section>
        </div>
      </section>
    </section>
  </div>
</body>
`;
